### PR TITLE
Suggestion/go doc prep

### PR DESCRIPTION
**Pull request by houdini91**
Draft PR to try and expose a cleaner go doc.
In this PR i focus on the `sbom` submodule.

 For reference our current `sbom` godoc can be found in https://pkg.go.dev/github.com/bom-squad/protobom@v0.3.0/pkg/sbom.
 
 The new version adds a bunch of descriptions to different the library API and data model.
 It also includes a set of description to the protobom document fields (i 

* `HashAlgorithmFromCDX` seems to be a duplicate of `HashAlgorithmFromCycloneDX`.
I added a deprecation message on `HashAlgorithmFromCycloneDX`.
* Added comments to the protobom it self, these comments propagate in to the golang generated file and godoc as well.
* Added some basic examples for adding roots, nodes, edges and metadata to the document directly.

> ANY Reviews and feedback would be greatly appreciated

## How to view the doc?

In order to view the Doc you can install godoc and then run
```
godoc -http=:8080
```
You can then view the godoc in the browser under `http://localhost:8080/pkg/github.com/bom-squad/protobom/pkg/sbom/`.

Attaching html for reference
[protobom_sbom_godoc.zip](https://github.com/bom-squad/protobom/files/13804351/protobom_sbom_godoc.zip)
 the godoc, note the way i share it only includes the plain text with out all the UI styles and css ..

### CODE REVIEW

**Comment by houdini91** [comment]
@puerco if you want to i think the PR is ready for review.

File: pkg/sbom/document_test.go [hunk 1]
```diff
@@ -1 +1,106 @@
 package sbom
+
```
Comment (puerco) [inline pkg/sbom/document_test.go:2]: TGhese examples are cool but there is a problem, they are in the sbom namespace. You should declare these in the `sbom_test` package to make sure they work when importing the library. 

Anyone following the examples should be using the functions with the package name like:

```golang
	// Create a new protobom document
	document := sbom.NewDocument()
```

As they are, the examples will fail when copying+pasting `document := NewDocument()` directly.

**Review by puerco — COMMENTED** [review]
Thanks for working on this @houdini91 , this is awesome. Here's an initial pass on the go code. I added a few comments but they mostly boil down to:

1. Let's mark the sbom functions that deal with the native formats as deprecated as that functionality should move to the native packages.
2. Ensure we standarize on the data structure names, eg `NodeList` as opposed to Node List.
3. The examples should be written from the point of view of someone using the library externally, move them into `sbom_test` to make sure they are called using the proper namespace.

Again great work

File: pkg/sbom/functions.go [hunk 2]
```diff
@@ -63,6 +64,7 @@ func NewNodeIdentifier(prefixes ...string) string {
 	return strings.Join(append(knownPrefixes, validPrefixes...), "-")
 }
 
+// EdgeTypeFromSPDX converts an SPDX2 edge type string to its corresponding edge type.
```
[the remark this answers is not in the document]
Comment (houdini91) [reply pkg/sbom/functions.go:67]: This one is also mapped to unserializer.EdgeTypeFromSPDX2?

File: pkg/sbom/node.go [hunk 3]
```diff
@@ -238,16 +239,16 @@ func (n *Node) Copy() *Node {
...
 func (n *Node) Equal(n2 *Node) bool {
 	if n2 == nil {
 		return false
 	}
 	return n.flatString() == n2.flatString()
 }
 
-// flatString returns a string representation of the node which can be used to
-// index the node or compare it against another
+// flatString returns a serialized representation of the node as a string,
```
[the remark this answers is not in the document]
Comment (houdini91) [reply pkg/sbom/node.go:250]: thanks, at some point i my brain switched was documentation everything in sight

**Comment by houdini91** [comment]
@puerco Thanks the detailed PR i appreciate it.
I think i took care of all your comments take another gander at the new version. 

The thing about the new deprecated functions is that do not yet find (or i simply missed it somehow) the alternative locations so i kinda guessed what they will be, can you refer and make sure i have deprecation message as you expect?

Any case this will fail our linting until we move our library to use the new underappreciated functions.

At the end of the PR i squash the commits so that we pass the DCO.

**Comment by puerco** [comment]
OK, sounds good, we can defer the deprecations to later when we replace the format-specific funcs with their native package equivalents :+1:  I'll try to work on it before our next meeting.

Be sure to squash and force push the branch as the DCO will not pass until you do.

**Comment by houdini91** [comment]
@puerco All tests pass, anything else needed?

File: pkg/sbom/functions.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (puerco) [reply pkg/sbom/functions.go:67]: The corresponding function should live in the SPDX2 unserializer

**Review by puerco — APPROVED** [review]
Awesome work Mikey! I see a few details here and there that could use some more polishing but with this PR the amount of documentation and examples are now improved substantially. I will merge it as it is and we can keep making it better incrementally as we go. Thanks @houdini91 !! :rocket: :rocket:

**Pull request merged**